### PR TITLE
[DoctrineBridge] Add support for doctrin/dbal v2.6 types

### PR DESCRIPTION
**Pull request by jvasseur**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

Add support for the following doctrine types in the property info extractor:
 - date_immutable
 - datetime_immutable
 - datetimetz_immutable
 - time_immutable
 - dateinterval

I didn't include the json type since it can be anything that can be stored in a json.

And add support for the dateinterval type for the form type guesser (the form component doesn't support using DateTimeImmutable instances).

### CODE REVIEW

**Comment by Hanmac** [comment]
i just wonder if the new types should add more  `DBALType::` constants too?

**Comment by jvasseur** [comment]
@Hanmac we can't use the constants for the new types if we want to stay compatible with `doctrine/dbal` < 2.6 where they don't exists.

**Comment by Hanmac** [comment]
@jvasseur ah i thought the constants where in our part too.
then i am okay with that.

we might use the constants when we drop the support for the older versions.

are the DBAL Type objects itself useful for this? i think no, but i am not sure

PS: in the Guesser should `'Symfony\Component\Form\Extension\Core\Type\CollectionType'` be replaced with `CollectionType::class` ?

**Comment by jvasseur** [comment]
The Type objects are not used by the guesser/property extractor, the decision is only done based on the name of the type.

I'm :-1: for using `::class` in the Guesser. It would mean that if we do that for all the types (only possible after 2.8 end of support) the class will have a lot of use at the start of the file just for this.

**Review by xabbuh — APPROVED** [review]

**Comment by stof** [comment]
The Form component supports reading immutable DateTime instances

**Comment by jvasseur** [comment]
@stof but it can't write back `DateTimeImmutable` instances meaning that you would have to do the conversion from a `DateTime` instance into a `DateTimeImmutable` in your setter.

I prefer to wait for proper support for `DateTimeImmutable` in the form component to add it to the guesser.

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php [hunk 1]
```diff
@@ -117,6 +117,15 @@ public function getTypes($class, $property, array $context = array())
                 case DBALType::TIME:
                     return array(new Type(Type::BUILTIN_TYPE_OBJECT, $nullable, 'DateTime'));
 
+                case 'date_immutable':
+                case 'datetime_immutable':
+                case 'datetimetz_immutable':
```
Comment (xabbuh) [inline src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php]: We don't have tests covering these three types, do we?

Comment (jvasseur) [inline src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php]: Only the `time` type was tested (and not the `date`, `datetime` and `datetimetz`) so I decided to do the same for the immutable variants.

File: src/Symfony/Bridge/Doctrine/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,10 @@
 CHANGELOG
 =========
 
+3.4.0
+-----
+ * added support for doctrine/dbal v2.6 types
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/CHANGELOG.md]: missing blank line just before

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @jvasseur.

**Pull request merged**